src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -436,6 +437,210 @@ void LoggingManagement::DestroyLoggingManagement(void)
     Platform::CriticalSectionExit();
 }
 
+/**
+ * Persist the Weave events of all importance types.
+ *
+ * Serializes the events in WeaveCircularTLVBuffer and the associated states into the
+ * supplied buffer and set mShutdownInProgress flag to true such that no further
+ * events can be added to the buffer.
+ *
+ * This method is intended to be used by devices that do not retain RAM while sleeping,
+ * allowing them to persist events before going to sleep and thereby prevent lost of
+ * events
+ */
+WEAVE_ERROR LoggingManagement::PersistEvent(uint8_t * buf, uint32_t bufSize, uint32_t & serializedLen)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    MarkShutdownInProgress();
+
+    Platform::CriticalSectionEnter();
+
+    TLVWriter writer;
+    writer.Init(buf, bufSize);
+
+    TLVType rootContainer;
+    err = writer.StartContainer(ProfileTag(kWeaveProfile_Common, kTag_SerializedEvent), kTLVType_Structure, rootContainer);
+    SuccessOrExit(err);
+
+    for (int importanceType = kImportanceType_First; importanceType < kImportanceType_Last + 1; importanceType++)
+    {
+        // Get the specified event buffer.
+        CircularEventBuffer * circular_event_buffer = GetImportanceBuffer((ImportanceType)importanceType);
+
+        TLVType container;
+        err = writer.StartContainer(ContextTag(importanceType), kTLVType_Structure, container);
```

Comment:
This should not use the important type as a tag.   In fact, the outer container should not be structure.  Instead, it should be an array of inner structures ("serialized event sets") with the importance expressed as a field within the inner structure.  E.g.:

```
serialized-event-state => ARRAY OF serialized-event-set

serialized-event-set => STRUCTURE
{
    importance-level [0]               : UNSIGNED INTEGER,
    event-data [1]                     : OCTET STRING,
    head-offset [2]                    : UNSIGNED INTEGER,
    length [3]                         : UNSIGNED INTEGER,
    first-event-id [4]                 : UNSIGNED INTEGER,
    last-event-id [5]                  : UNSIGNED INTEGER,
    first-event-timestamp [6]          : UNSIGNED INTEGER,
    last-event-timestamp [7]           : UNSIGNED INTEGER,
    first-event-utc-timestamp [8,opt]  : UNSIGNED INTEGER,
    first-event-utc-timestamp [9,opt]  : UNSIGNED INTEGER,
    last-event-utc-timestamp [10,opt]  : UNSIGNED INTEGER,
}
```